Initial commit of LED colour adjuster project

aela-oroa opened the pull request:
This commit includes the main application files for controlling LED colours via serial communication. It sets up the basic structure of the project and includes functionality for adjusting brightness, power, and colour.

aela-oroa committed:
Initial commit of LED colour adjuster project

This commit includes the main application files for controlling LED colours via serial communication. It sets up the basic structure of the project and includes functionality for adjusting brightness, power, and colour.

bohdan-moiseienko commented on mainwindow.cpp:
```diff
@@ -0,0 +1,197 @@
+}
+
+// Create command for serial communication
+QString MainWindow::createCommand(const QString &code, int value)
```
The first argument can also be passed as `int`. And please rename it to "address"

aela-oroa committed:
Refactor createCommand to use int address

Replaced QString code with int address for creating commands, updated hex conversion logic. Updated all function calls.

aela-oroa committed:
Refactor code, update UI, and add LED selection feature

aela-oroa committed:
Update README.md

bohdan-moiseienko commented on led-adjust.pro:
```diff
@@ -0,0 +1,25 @@
+# In order to do so, uncomment the following line.
+#DEFINES += QT_DISABLE_DEPRECATED_BEFORE=0x060000    # disables all the APIs deprecated before Qt 6.0.0
+
+SOURCES += \
+    main.cpp \
+    mainwindow.cpp
+
+HEADERS += \
+    mainwindow.h
+
+FORMS += \
+    mainwindow.ui
```
Please add all new files you added to the project here. Currently it cant be built with this .pro file.

aela-oroa replied:
Oops, I forgot to add the new version of this file. It's fixed now.

bohdan-moiseienko commented on mainwindow.h:
```diff
@@ -0,0 +1,37 @@
+
+private:
+    Ui::MainWindow *ui;
+    BoardWithLeds *board;
```
Do we need to create the object on the heap? Can we just store it like an instance variable like this: 
`BoardWithLeds board;`

aela-oroa replied:
I initially used the heap due to some issues with the LEDs, but the instance variable works fine now.

bohdan-moiseienko commented on boardwithleds.h:
```diff
@@ -0,0 +1,29 @@
+class BoardWithLeds
+{
+public:
+    BoardWithLeds(Led led);
```
If your constructor takes only one argument it is better to mark it `explicit` in order to avoid possibility of implicit type cast

bohdan-moiseienko commented on portcontrolwidget.h:
```diff
@@ -0,0 +1,38 @@
+    void portsAvailable(bool available);
+
+private slots:
+    void on_connectWidget_clicked();
```
make the naming of your methods and variables consistent: `onConnectWidgetClicked`

aela-oroa replied:
I get the consistency thing, but `on_connectWidget_clicked` is just how Qt Designer names slots. If I change them in the .h and .cpp files, it breaks the UI since Qt can't find them anymore.

Any tips on doing that without breaking the UI?..

aela-oroa replied:
Any tips on doing that without breaking the UI?

bohdan-moiseienko replied:
You can create the connection manually in the code.

aela-oroa replied:
Should I rename all slots and connect them manually? Won’t that just add redundant code?

bohdan-moiseienko replied:
Yes, exactly. It's a common practice and it's the only way to get rid of autogenerated names.

bohdan-moiseienko commented on mainwindow.cpp:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    // Instantiate new board for selected LED and update PortControlWidget
+    board = new BoardWithLeds(selectedLed);
```
Why do we need a new board object each time we switch leds? Can we avoid this?

aela-oroa replied:
Yeah, I think we can avoid creating a new board object each time by adding a public method to switch LEDs.

bohdan-moiseienko commented on portconnectwidget.cpp:
this file looks odd here, do we need it?

aela-oroa replied:
Guess it was from when I messed up creating the custom widget class.

aela-oroa committed:
Remove mistakenly added portconnectwidget.cpp

aela-oroa committed:
Update led-adjust.pro

aela-oroa committed:
Refactor BoardWithLeds initialization, improve naming consistency, and add resetLedStates method

Refactored BoardWithLeds to avoid creating a new object for each LED change, improved naming consistency, marked constructor explicit, and added resetLedStates method for resetting LEDs.

aela-oroa committed:
Add circuit design file
